### PR TITLE
AWS database access configurator

### LINKED ISSUE #8974: Implement database access IAM configurator
### What

Implement `teleport db configure aws` command described in [RFD46](https://github.com/gravitational/teleport/blob/master/rfd/0046-database-access-config.md).

### PR DESCRIPTION
**Pull request by gabrielcorado**
Database configure bootstrap implementation from [RFD46](https://github.com/gravitational/teleport/blob/master/rfd/0046-database-access-config.md).

Some notes for reviewers:
* The interaction with AWS, data validation, and policy generation is implemented inside the `lib/configurators/databases` package;
* Policies were taken from [the documentation page](https://goteleport.com/docs/ver/8.0/database-access/guides/rds/) and tested;
* Documentation updates will be done in a separated PR;

### CODE REVIEW

File: lib/configurators/databases/aws.go [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+
+	//
+	policyArn, versions, err := a.retrieveCurrentPolicy(documentName)
+	if err != nil && !trace.IsNotFound(err) {
+		return "", trace.Wrap(err)
+	}
+
+	// If no versions were found, we need to create a new policy.
+	if trace.IsNotFound(err) {
+		resp, err := iamClient.CreatePolicyWithContext(context.Background(), &iam.CreatePolicyInput{
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:273]: Let's pass the context all the way from the caller. Same throughout.

File: lib/configurators/databases/aws.go [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+		return "", nil, trace.NotFound("AWS identity not provided, unable to determine policy account ID")
+	}
+
+	policyArn := fmt.Sprintf("arn:aws:iam::%s:policy/%s", a.awsIdentity.GetAccountID(), policyName)
+	client, err := a.getAWSIAMClient()
+	if err != nil {
+		return "", nil, trace.Wrap(err)
+	}
+
+	getPolicyResp, err := client.GetPolicy(&iam.GetPolicyInput{PolicyArn: aws.String(policyArn)})
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:230]: Let's use XXXWithContext SDK methods and pass context object everywhere from the caller.

File: lib/configurators/databases/aws.go [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	iamClient, err := a.getAWSIAMClient()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	//
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:265]: Unfinished comment?

File: lib/configurators/databases/aws.go [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+	return policyArn, resp.Versions, nil
+}
+
+// upsertPolicyDocument creates a new Policy or creates a Policy version if a
+// policy with the same name already exists.
+//
+// Since policies can hold only five versions, we need to delete a policy
+// version (if the limit is reached) and create a new version.
+// Ref: https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-versioning.html#version-limits
+func (a *awsConfigurator) upsertPolicyDocument(documentName, documentDescription string, document *awslib.PolicyDocument) (string, error) {
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:254]: Feels like this method should not be a part of configurator but rather some helper method that accepts AWS client and other parameters or maybe some a method of AWSClient object that wraps AWS clients and provides helper methods like this one.

File: lib/configurators/databases/aws.go [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+		})
+		if err != nil {
+			return "", wrapAWSError(err)
+		}
+
+		return *resp.Policy.Arn, nil
+	}
+
+	// Check number of policy versions and delete one if necessary.
+	if len(versions) == 5 {
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:289]: `5` should probably be a constant somewhere.

File: lib/configurators/databases/aws.go [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+		// Sort versions based on create date.
+		sort.Slice(versions, func(i, j int) bool {
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:291]: Can this logic be factored out into `DeleteOldestPolicy` or something similar e.g. on that AWSClient wrapper I suggested above.

File: lib/configurators/databases/aws.go [hunk 7]
```diff
@@ -0,0 +1,656 @@
...
+	if a.awsSession != nil {
+		return a.awsSession, nil
+	}
+
+	var err error
+	a.awsSession, err = awssession.NewSessionWithOptions(awssession.Options{
+		// TODO: remove this, only used to test locally
+		// Config: aws.Config{
+		// 	Endpoint: aws.String("http://localhost:4566"),
+		// },
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:418]: Don't forget to remove this.

File: lib/configurators/databases/aws.go [hunk 8]
```diff
@@ -0,0 +1,656 @@
...
+	default:
+		return "", targetTypeUnknown, trace.BadParameter("not able to identify the target type")
+	}
+
+	return target, targetType, nil
+}
+
+// buildPolicyDocuments generates policy and policy boundary document based on
+// flags and fileConfig.
+func buildPolicyDocuments(targetType targetType, flags BootstrapFlags, fileConfig *config.FileConfig, identity awslib.Identity) (*awslib.PolicyDocument, *awslib.PolicyDocument, error) {
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:459]: I think there's too much going on in this method, it's pretty hard to follow.

First, I would split it in 2: separate for generating policy document and boundary document. This will align pretty well with my suggestion above about having separate "steps".

Also, I would refactor it so the flow/logic is more clear, for example:

```go
func buildPolicyDocument(...) (*awslib.PolicyDocument, error) {
  var statements []xxx
  if rdsEnabled() {
    rdsStatements, err := getRDSStatements()
    // check err
    statements = append(statements, rdsStatements...)
  }
  if auroraEnabled() {
    // getAuroraStatements()
  }
  if redshiftEnabled() {
    // getRedshiftStatements()
  }
  return awslib.NewPolicyDocument(statements)
}

func buildBoundaryDocument(...) (*awslib.PolicyDocument, error) {
  // similar to above
}
```

File: lib/configurators/databases/aws.go [hunk 9]
```diff
@@ -0,0 +1,656 @@
...
+
+	return []*awslib.Statement{
+		{Effect: awslib.EffectAllow, Actions: []string{"redshift:GetClusterCredentials"}, Resources: resources},
+		{Effect: awslib.EffectAllow, Actions: []string{"redshift:DescribeClusters"}, Resources: []string{"*"}},
+	}, nil
+}
+
+// formatPolicyDocument formats the PolicyDocument in a "friendly" format, which
+// can be presented to end users.
+func formatPolicyDocument(policy *awslib.PolicyDocument) (string, error) {
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:589]: Can this be a `Marshal` method on the PolicyDocument?

File: lib/configurators/databases/aws.go [hunk 10]
```diff
@@ -0,0 +1,656 @@
...
+	b, err := json.MarshalIndent(policy, "", "    ")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return string(b), nil
+}
+
+// encodePolicyDocument encode PolicyDocument into JSON.
+func encodePolicyDocument(policy *awslib.PolicyDocument) (string, error) {
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:599]: Is this needed? Can we just always marshal it with indent so we can use the same method for both presenting to users and sending via API?

File: lib/configurators/databases/aws.go [hunk 11]
```diff
@@ -0,0 +1,656 @@
...
+		if *tag.Key == policyTeleportTagKey {
+			return true
+		}
+	}
+
+	return false
+}
+
+// wrapAWSError wraps an AWS error accordingly.
+func wrapAWSError(err error) error {
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:620]: Is this method needed? Why not just call awslib.ConvertAWSError(e)?

File: lib/configurators/databases/aws.go [hunk 12]
```diff
@@ -0,0 +1,656 @@
...
+	switch e := err.(type) {
+	case awserr.RequestFailure:
+		return awslib.ConvertAWSRequestFailureError(e)
+	default:
+		return trace.Wrap(err)
+	}
+}
+
+// targetType represents types that will have the policies attached to.
+type targetType int
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:630]: Can this be a string? Then you won't need String() method.

I would also consider moving this to lib/cloud/aws/identity.go and calling it IdentityType or something.

File: lib/configurators/databases/databases.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	Manual bool
+	// PolicyName name of the generated policy.
+	PolicyName string
+	// AttachToUser user that the generated policies will be attached to.
+	AttachToUser string
+	// AttachToRole role that the generated policies will be attached to.
+	AttachToRole string
+	// IncludeRedshiftWildcard indicates if Redshift permissions must be
+	// included using a wildcard.
+	IncludeRedshiftWildcard bool
```
Comment (r0mant) [inline lib/configurators/databases/databases.go:41]: I'm not sure what this flag is for?

File: lib/configurators/databases/databases.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	// ForceRDSPermissions forces the presence of RDS permissions.
+	ForceRDSPermissions bool
+	// ForceAuroraPermissions forces the presence of Redshift permissions.
+	ForceRedshiftPermissions bool
```
Comment (r0mant) [inline lib/configurators/databases/databases.go:45]: Are these needed? Can we simplify and just determine permissions based on the config file alone for now? Would probably simplify a few things.

File: lib/configurators/databases/databases.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+// ExecutionResult execution step result.
+type ExecutionResult struct {
+	// Step description of the current step.
+	Description string
+	// Err error returned by the step.
+	Err error
+}
+
+// Configurator
```
Comment (r0mant) [inline lib/configurators/databases/databases.go:66]: Unfinished comment.

File: tool/teleport/common/teleport.go [hunk 16]
```diff
@@ -205,6 +215,35 @@ func Run(options Options) (app *kingpin.Application, executedCommand string, con
 	dbStartCmd.Flag("diag-addr", "Start diagnostic prometheus and healthz endpoint.").StringVar(&ccf.DiagnosticAddr)
 	dbStartCmd.Flag("insecure", "Insecure mode disables certificate validation").BoolVar(&ccf.InsecureMode)
 	dbStartCmd.Alias(dbUsageExamples) // We're using "alias" section to display usage examples.
+	dbConfigure := dbCmd.Command("configure", "Database agent bootstrap.")
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:218]:
```suggestion
	dbConfigure := dbCmd.Command("configure", "Bootstraps database service configuration and cloud permissions.")
```

File: tool/teleport/common/teleport.go [hunk 17]
```diff
@@ -490,6 +535,255 @@ func onForward() error {
...
+	}
+
+	return nil
+}
+
+// requestConfirmation waits for user confirmation and returns `true` if the
+// user confirms. The confirmation is done by typing "yes" or "y". The caller
+// can provide an autoConfirm param which will skip the user input and return
+// `true`.
+func requestConfirmation(autoConfirm bool) (bool, error) {
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:767]: We have lib/utils/prompt/confirmation.go.

File: tool/teleport/common/teleport.go [hunk 18]
```diff
@@ -490,6 +535,255 @@ func onForward() error {
...
+	return nil
+}
+
+// configureDatabaseAWSPrintFlags flags of the "db configure aws print-iam"
+// subcommand.
+type configureDatabaseAWSPrintFlags struct {
+	configureDatabaseAWSFlags
+	//
+	policyOnly bool
+	//
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:663]: Unfinished comments.

File: tool/teleport/common/teleport.go [hunk 19]
```diff
@@ -490,6 +535,255 @@ func onForward() error {
...
+
+	if f.types == "" {
+		return trace.BadParameter("at least one --type should be provided: %s", strings.Join(awsDatabaseTypes, ","))
+	}
+
+	f.typesList = strings.Split(f.types, ",")
+typesListLoop:
+	for _, dbType := range f.typesList {
+		for _, supportedDBType := range awsDatabaseTypes {
+			if dbType == supportedDBType {
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:646]: If you invert this to `!=`, you won't need the loop label.

**Review by r0mant — COMMENTED** [review]
@gabrielcorado Lots of work, nice job! I think all the main pieces are there but I left a few refactoring suggestions which I think will improve the structure and readability of the implementation.

File: lib/configurators/databases/aws.go [hunk 20]
```diff
@@ -0,0 +1,656 @@
...
+
+func (a *awsConfigurator) Instructions() []Instruction {
+	return []Instruction{
+		{Description: fmt.Sprintf("1. Create IAM policy %q:", a.policyName), Details: a.formattedPolicyDocument},
+		{Description: fmt.Sprintf("2. Create IAM boundary policy %q:", a.policyBoundaryName), Details: a.formattedPolicyBoundaryDocument},
+		{Description: fmt.Sprintf("3. Attach policies to %s %q.", a.targetType.String(), a.target)},
+	}
+}
+
+func (a *awsConfigurator) Execute(ctx context.Context) []ExecutionResult {
```
[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/configurators/databases/aws.go:136]: 👍  I liked this kind of builder pattern idea.

File: lib/configurators/databases/databases.go (hunk 14 above)
Comment (gabrielcorado) [reply lib/configurators/databases/databases.go:45]: We need them to use the configurator for the `aws create-iam` and `aws print-iam` commands. Those commands define if the policy should have RDS/Redshift permissions using flags (instead of the config file).

**Review by smallinsky — COMMENTED** [review]
Left some comments. LGTM when remaining comments will be addressed.

**Comment by gabrielcorado** [comment]
@r0mant and @smallinsky, thanks for the review; based on your comments, I've made a few changes:
* Moved `UpsertPolicy`, `RetrievePolicy`, `AttachPolicy` and `AttachPolicyBoundary` to the `lib/cloud/aws` package since they were not specific to the configurator;
* Changed the `Configurator` to return a list of `Actions` where each action can be executed (and tested) independently;
* Split the building policy into smaller functions, making it easier to compose and test;

File: lib/configurators/databases/aws.go (hunk 11 above)
Comment (gabrielcorado) [reply lib/configurators/databases/aws.go]: It was needed since the `ConvertAWSError` expected an `awserr.RequestFailure`. I've updated the function to receive an `error` and convert internally, so we don't need this function anymore.

File: lib/configurators/databases/aws.go (hunk 12 above)
Comment (gabrielcorado) [reply lib/configurators/databases/aws.go]: This type is no longer necessary, and the configurator now relies on the `Identity` type.

File: lib/configurators/databases/databases.go (hunk 13 above)
Comment (gabrielcorado) [reply lib/configurators/databases/databases.go:41]: It was left after a refactoring.

File: tool/teleport/common/teleport.go (hunk 19 above)
Comment (gabrielcorado) [reply tool/teleport/common/teleport.go:708]: We need the outer loop label because the return statement is on it. We still need to `continue` the outer loop, even changing the condition. The only alternative I see here to make things easier to understand is to place in the inner loop in a separated function so that we could have:
```go
for _, dbType := range f.typesList {
	if !isSupported(dbType) {
		return trace.BadParameter("...")
	}
}
```

What do you think?

File: tool/teleport/common/teleport.go [hunk 21]
```diff
@@ -490,6 +535,255 @@ func onForward() error {
...
+	instructions := configurator.Instructions()
+	if flags.policyOnly {
+		// Policy is present at the details of the first instruction.
+		fmt.Println(instructions[0].Details)
+		return nil
+	}
+
+	if flags.boundaryOnly {
+		// Policy boundary is present at the details of the second instruction.
+		fmt.Println(instructions[1].Details)
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply tool/teleport/common/teleport.go]: I've added an `IsEmpty` check, which will guarantee those slice items are present.

File: lib/configurators/databases/aws.go [hunk 22]
```diff
@@ -0,0 +1,639 @@
...
+			}
+		}
+	}
+
+	if len(resources) == 0 {
+		return []*awslib.Statement{}, nil
+	}
+
+	return []*awslib.Statement{
+		{Effect: awslib.EffectAllow, Actions: []string{"redshift:GetClusterCredentials"}, Resources: resources},
```
Comment (r0mant) [inline lib/configurators/databases/aws.go:625]: Hmm, do we need to include this in the policy? This is what Teleport agent creates itself during auto-IAM configuration, no? And the `resources` loop above probably isn't needed also.

File: tool/teleport/common/teleport.go (hunk 19 above)
Comment (r0mant) [reply tool/teleport/common/teleport.go:708]: I have a better proposal - we already have `utils.SliceContainsStr` - can we use that instead of implementing double-loop?

File: tool/teleport/common/teleport.go [hunk 23]
```diff
@@ -578,6 +625,257 @@ func onForward() error {
...
+	}
+
+	return nil
+}
+
+// configureDatabaseAWSPrintFlags flags of the "db configure aws print-iam"
+// subcommand.
+type configureDatabaseAWSPrintFlags struct {
+	configureDatabaseAWSFlags
+	//
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:723]: Missing comments here and below.

**Review by r0mant — COMMENTED** [review]
Much cleaner now :+1: I have left a few more suggestions.

**Review by r0mant — COMMENTED** [review]
@gabrielcorado Also, have you tested this on an actual AWS node with IAM role attached? Let's please test this scenario.

File: lib/configurators/databases/aws.go (hunk 22 above)
Comment (gabrielcorado) [reply lib/configurators/databases/aws.go:625]: You're right. This is similar to the one Teleport generates. We still need to place them on the boundary policy, or Teleport won't have enough permissions to attach to its user/role. And this loop is reducing the number of "wildcard permissions" we could place a default wildcard permission (like the one added when Redshift auto-discovery is enabled).

File: lib/configurators/databases/aws.go [hunk 24]
```diff
@@ -0,0 +1,639 @@
...
+
+	formattedPolicy, err := formatPolicyDocument(policyDocument)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	formattedPolicyBoundary, err := formatPolicyDocument(policyBoundaryDocument)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/configurators/databases/aws.go:301]: It is kept outside so that neither we need an `awsPolicyCreator/awsPoliciesAttacher` constructors nor the `Details` function needs to return an error. For me, it makes sense to keep this close to the policy build since if there is an issue with it, the root cause is "close by".

File: lib/configurators/databases/aws.go [hunk 25]
```diff
@@ -0,0 +1,639 @@
...
+func (a *awsPoliciesAttacher) Details() string {
+	return ""
+}
+
+// Execute retrieves policy and policy boundary ARNs from `context.Context` and
+// attach them to the `target`.
+func (a *awsPoliciesAttacher) Execute(ctx context.Context) (context.Context, error) {
+	if a.helper == nil {
+		return ctx, trace.BadParameter("policy helper not initialized")
+	}
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/configurators/databases/aws.go:198]: This check is necessary if the caller tries to execute an Action generated in "manual" mode (meaning it won't have the `Policies` initialized). We could have different structs for when it's manual or not.

File: lib/configurators/databases/aws.go [hunk 26]
```diff
@@ -0,0 +1,639 @@
...
+// empty string.
+func (a *awsPoliciesAttacher) Details() string {
+	return ""
+}
+
+// Execute retrieves policy and policy boundary ARNs from `context.Context` and
+// attach them to the `target`.
+func (a *awsPoliciesAttacher) Execute(ctx context.Context) (context.Context, error) {
+	if a.helper == nil {
+		return ctx, trace.BadParameter("policy helper not initialized")
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/configurators/databases/aws.go]: I've changed this logic to no longer use the `context.Context` as Roman suggested.

**Review by smallinsky — APPROVED** [review]

File: tool/teleport/common/teleport.go [hunk 27]
```diff
@@ -593,6 +641,252 @@ func onCheckHome() error {
...
 
+// configureDatabaseBootstrapFlags database configure bootstrap flags.
+type configureDatabaseBootstrapFlags struct {
+	config  dbconfigurators.BootstrapFlags
+	confirm bool
+}
+
+// onConfigureDatabaseBootstrap subcommand that bootstraps configuration for
+// database agents.
+func onConfigureDatabaseBootstrap(flags configureDatabaseBootstrapFlags) error {
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:652]: Can we please move all these new methods out of `teleport.go` and into its own source file e.g. `configurator.go`?

File: lib/cloud/aws/policy.go [hunk 28]
```diff
@@ -128,11 +158,236 @@ func (p *PolicyDocument) Delete(effect, action, resource string) {
 	p.Statements = statements
 }
 
+// PolicyHelper set of IAM Policy helper functions defined as an interface to
```
Comment (r0mant) [inline lib/cloud/aws/policy.go:161]:
```suggestion
// Policies set of IAM Policy helper functions defined as an interface to
```

**Review by r0mant — COMMENTED** [review]
A final batch of comments, should be good to go after that.

@gabrielcorado Also, have you seen my comment about trying this out on an actual EC2 node with an IAM role attached? Have you tested this scenario? Let me know if I can help you with it.

File: lib/configurators/databases/aws.go (hunk 22 above)
Comment (gabrielcorado) [reply lib/configurators/databases/aws.go:568]: > Hmm does that mean though that they will have to re-run configurator after adding each new database?
For static Redshift database, yes. It is the tradeoff of having a more restrictive boundary.

We can identify that the agent has dynamic registration and go for the wildcard in those cases since we don't know the complete databases list. And since it is the recommended way, going with only wildcard permissions for Redshift sounds ok.

File: tool/teleport/common/teleport.go [hunk 29]
```diff
@@ -593,6 +602,12 @@ func onCheckHome() error {
 	return nil
 }
 
+// configureDatabaseBootstrapFlags database configure bootstrap flags.
+type configureDatabaseBootstrapFlags struct {
+	config  dbconfigurators.BootstrapFlags
+	confirm bool
+}
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:609]: Should probably move this to `configurator.go` too?

File: tool/teleport/common/configurator.go [hunk 30]
```diff
@@ -0,0 +1,307 @@
...
+
+// executeDBConfiguratorActions iterate over all actions, executing and priting
+// their results.
+func executeDBConfiguratorActions(ctx context.Context, configuratorName string, actions []dbconfigurators.ConfiguratorAction) error {
+	actionContext := &dbconfigurators.ConfiguratorActionContext{}
+	for _, action := range actions {
+		err := action.Execute(ctx, actionContext)
+		printDBBootstrapActionResult(configuratorName, action, err)
+		if err != nil {
+			return err
```
Comment (r0mant) [inline tool/teleport/common/configurator.go:286]:
```suggestion
			return trace.Wrap(err)
```

File: lib/configurators/databases/aws.go (hunk 22 above)
Comment (r0mant) [reply lib/configurators/databases/aws.go:567]: Yes, let's do that. Do you need to update this method then to remove the loop above?

**Review by r0mant — APPROVED** [review]

**Comment by gabrielcorado** [comment]
@r0mant Yeah, I did the testing with roles attached to EC2 instances to access RDS and Redshift databases.
